subsys/net/ip/l2/wifi_shell.c

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief WiFi shell module
+ */
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <shell/shell.h>
+#include <misc/printk.h>
+#include <init.h>
+
+#include <net/net_if.h>
+#include <net/wifi_mgmt.h>
+#include <net/net_event.h>
+
+#define WIFI_SHELL_MODULE "wifi"
+
+#define WIFI_SHELL_MGMT_EVENTS (NET_EVENT_WIFI_SCAN_RESULT |		\
+				NET_EVENT_WIFI_SCAN_DONE |		\
+				NET_EVENT_WIFI_CONNECT_RESULT |		\
+				NET_EVENT_WIFI_DISCONNECT_RESULT)
+
+static struct net_mgmt_event_callback wifi_shell_mgmt_cb;
+
+static void handle_wifi_scan_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_scan_result *entry =
+		(const struct wifi_scan_result *)cb->info;
+
+	printk("SSID: %s (%u) - Channel %u - Security: %s - RSSI: %d\n",
+	       entry->ssid, entry->ssid_length, entry->channel,
+	       entry->security == WIFI_SECURITY_TYPE_PSK ?
+	       "WPA/WPA2" : "OPEN", entry->rssi);
+}
+
+static void handle_wifi_scan_done(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *)cb->info;
+
+	if (status->status) {
+		printk("Scan request failed (%d)\n", status->status);
+	} else {
+		printk("Scan request done\n");
+	}
+}
+
+static void handle_wifi_connect_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *) cb->info;
+
+	if (status->status) {
+		printk("Connection request failed (%d)\n", status->status);
+	} else {
+		printk("Connection request done\n");
+	}
+}
+
+static void handle_wifi_disconnect_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *) cb->info;
+
+	if (status->status) {
+		printk("Disconnection request failed (%d)\n", status->status);
```

Comment:
While testing I found that a disconnect event can occur without a disconnect request, for example if the AP disconnects the device via an error or removing a hotspot on a mobile phone.
I found it can also occur during a connect after a previous connect to the same SSID (i.e., the AP was disconnecting, then starting the reconnect).
So, perhaps we just print here "Disconnected".

Reply:
Yep, forgot that obvious use case when the AP dies, or becomes too far etc etc...